src/modm/driver/usb/stusb4500.hpp

```diff
@@ -0,0 +1,160 @@
+		DpmSnkPdo1 = 0x85,
+		DpmSnkPdo2 = 0x89,
+		DpmSnkPdo3 = 0x8D,
+	};
+
+public:
+	struct RdoRegStatusData {
+		uint32_t MaxCurrent; /// [mA]
+		uint32_t OperatingCurrent; /// [mA]
+		uint8_t ObjectPos;
+	};
+};
+
+/**
+ * @ingroup modm_driver_stusb4500
+ * @author Henrik Hose, Raphael Lehmann
+ */
+template<class I2cMaster>
+class Stusb4500 : public stusb4500, public modm::I2cDevice< I2cMaster, 4 >
+{
+public:
+	inline Stusb4500(uint8_t address=0x28):
+		modm::I2cDevice<I2cMaster,4>(address)
+	{};
+
+	/**
+	 * Configures a PDO.
+	 *
+	 * Note: According to the USB standard the first PDO (1) always has to set
+	 *  voltage to 5000 [mV].
+	 *
+	 * @param	pdoNumber Which PDO to configure, range 1 to 3.
+	 * @param	voltage	  in mV
+	 * @param	current	  in mA
+	 */
+	inline modm::ResumableResult<bool>
+	configurePdo(uint8_t pdoNumber, uint32_t voltage, uint32_t current)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_CALL(readRegister<4>(DpmSnkPdos.at(pdoNumber - 1), buffer2.data()));
+
+		current /= 10;
+		voltage /= 50;
+
+		buffer2.at(0) = current;
+		buffer2.at(1) = ((current >> 8) & 0b11) | ((voltage & 0b11'1111 ) << 2);
+		buffer2.at(2) = (buffer2.at(2) & 0b1111'0000) | ((voltage >> 6) & 0b1111);
+
+		RF_END_RETURN_CALL(updateRegister<4>(DpmSnkPdos.at(pdoNumber - 1), buffer2.data()));
+	}
+
+	/**
+	 * Sets the valid PDO.
+	 *
+	 * @param	pdoNumber Which PDO to set valid, range 1 to 3.
+	 */
+	modm::ResumableResult<bool>
+	setValidPdo(uint8_t pdoNumber)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_END_RETURN_CALL(updateRegister<1>(Register::DpmPdoNumb, &pdoNumber));
+	}
+
+	/**
+	 * Retrieve RDO status from STUSB4500 chip.
+	 *
+	 * @return	RdoRegStatusData object. Typically only
+	 *  RdoRegStatusData::MaxCurrent [mA] is of interest.
+	 */
+	inline modm::ResumableResult<RdoRegStatusData>
+	getRdoRegStatus()
+	{
+		RF_BEGIN();
+
+		RF_CALL(readRegister<4>(Register::RdoRegStatus, buffer2.data()));
+
+		// MaxCurrent: Bits 9..0; OperatingCurrent: Bits 19..10; ObjectPos: Bits 30..28
+		rdoRegStatusData.MaxCurrent = buffer2.at(0) | ((buffer2.at(1) & 0b11) << 8);
+		rdoRegStatusData.MaxCurrent *= 10;
+		rdoRegStatusData.OperatingCurrent = ((buffer2.at(1) & 0b1111'1100) >> 2) | ((buffer2.at(2) & 0b1111) << 6);
+		rdoRegStatusData.OperatingCurrent *= 10;
+		rdoRegStatusData.ObjectPos = (buffer2.at(3) & 0b0110'0000) >> 5;
+
+		RF_END_RETURN(rdoRegStatusData);
+	}
+
+private:
+	template<uint32_t length>
+	modm::ResumableResult<bool>
+	readRegister(Register reg, uint8_t* output)
+	{
+		RF_BEGIN();
+		buffer[0] = uint8_t(reg);
+		this->transaction.configureWriteRead(buffer.data(), 1, output, length);
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+	template<uint32_t length>
+	modm::ResumableResult<bool>
+	updateRegister(Register reg, uint8_t* data)
+	{
+		static_assert(length <= (std::tuple_size<decltype(buffer)>::value - 1), "buffer too small");
```

Comment:
Pass length as an argument and use `length = std::min(length, 5);`

Reply:
A span would be appropriate. And why use tuple_size. std::array is constexpr.

`static_assert(4 <= (buffer.size() - 1), "buffer too small");`

Reply:
Nope. `array.size()` does not work in static_assert() because `buffer` is not constexpr. Try it out yourself...

Reply:
> use `length = std::min(length, 5);`

No, because I don't like silent failures. Run-time checking the `length` parameter and abort with `return false` now.